### PR TITLE
fix: lint files and create a global cache

### LINKED ISSUE #75: All modules are linted in watch mode after changing only a single file
- Operating System: macOS 11.1
- Node Version: 15.7.0
- NPM Version: yarn berry PnP
- webpack Version: 4.46.0
- eslint-webpack-plugin Version: 2.5.0

### Expected Behavior

Only lint modules that were rebuilt after changing a file.

### Actual Behavior

Every module is linted after changing a file.

### Code

```js
new ESLintPlugin({
  extensions: ['js', 'vue'],
})
```

### How Do We Reproduce?

It seems there was a recent change to which hooks were used to lint the files. Before `succeedModule` was used which was only called with the rebuilt modules after a file change, but now `finishModules` is used which passes every module in the compilation, so all files get linted.

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Linting the necessary files and create a more intelligent cache of the result

Resolves #75 , #74 

### Additional Info

Thanks @watjurk

### CODE REVIEW

**Comment by ricardogobbosouza** [comment]
/cc @alexander-akait 

Can use the [invalid](https://github.com/webpack-contrib/eslint-webpack-plugin/pull/79/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R91) hook to invalidate cache?
Apparently it worked well

WDTY?

**Comment by alexander-akait** [comment]
`invalid` for watching, better don't use this hook, it will have same bad side effects with plugins and watching/serve

**Comment by ricardogobbosouza** [comment]
Can we keep it that way in v2 and fix it in v3? If not, we will have performance problems!

Or do you have any idea how to do this?

**Comment by alexander-akait** [comment]
I need to look at this issue, I would prefer to return to it in 3 versions, is it not critical now?

**Comment by watjurk** [comment]
Unfortunately I think this is critical. 
> it will have same bad side effects with plugins and watching/serve

what side effects ?

**Comment by alexander-akait** [comment]
Because some projects use manually `invalidate` for `serve`, need investigate how it can break this plugin

**Comment by watjurk** [comment]
@alexander-akait could you provide more details on how this can break this plugin, I'll be happy to investigate.
For issue #83 i would suggest using both compilation.hooks.finishModules and compilation.hooks.succeedModule in following configuration:

```js
const seen = new Set()

compilation.hooks.finishModules.tap(ESLINT_PLUGIN, (modules) => {
  /** @type {string[]} */
  const files = [];

  // @ts-ignore
  for (const { resource } of modules) {
    if (resource) {
      const [file] = resource.split('?');

      if (
        file &&
        !seen.has(file) &&
        isMatch(file, wanted) &&
        !isMatch(file, exclude)
      ) {
        seen.add(file)
        files.push(file);
      }
    }
  }

  if (files.length > 0) {
    lint(files);
  }
});

compilation.hooks.succeedModule.tap(ESLINT_PLUGIN, (module) => {
  const { resource } = module
  const [file] = resource.split('?'); //? maybe move this to function
  if (
    file &&
    !seen.has(file) &&
    isMatch(file, wanted) &&
    !isMatch(file, exclude)
  ) {
    seen.add(file)
    lint(file)
  }
})
```

"Why don't use only compilation.hooks.succeedModule" ?
Because then (in watch mode) we would only lint files that have changed. 
"Yes but then we could use crossRunResultStorage". 
Then we would again face issue #74.

File: src/linter.js [hunk 1]
```diff
@@ -39,8 +38,6 @@ export default function linter(key, options, compilation) {
   /** @type {Promise<LintResult[]>[]} */
   const rawResults = [];
 
-  const crossRunResultStorage = getResultStorage(compilation);
```
Comment (jsg2021) [inline src/linter.js:42]: This cache held all results that had been computed... invalidations were supposed to be deleted from this and re-run.. so we don't recompute results for files that did not change... but still show them if they had issues.

**Comment by jsg2021** [comment]
This feels like reinventing the results storage that was removed when all that was needed was to remove results for invalidated entries.

File: src/linter.js (hunk 1 above)
Comment (jsg2021) [reply src/linter.js:42]: Related #83

**Comment by jsg2021** [comment]
> "Yes but then we could use crossRunResultStorage."
> Then we would again face issue #74.

Just delete entries from it on invalidation events.

**Comment by watjurk** [comment]
@ricardogobbosouza can we get https://github.com/webpack-contrib/eslint-webpack-plugin/pull/79#issuecomment-778366981 into fix-lint-and-cache branch ?
@alexander-akait could you provide some details on how compiler.invalid could break this PR ?

**Pull request closed without merging**